pulsar-client-cpp/python/src/config.cc

```diff
@@ -110,21 +112,29 @@ class LoggerWrapper: public Logger {
 
    public:
 
-    LoggerWrapper(const std::string &logger, PyObject* pyLogger) {
+    LoggerWrapper(const std::string &filename, PyObject* pyLogger) {
         _pyLogger = pyLogger;
         Py_XINCREF(_pyLogger);
 
+        fallbackLogger = (new ConsoleLoggerFactory())->getLogger(filename);
```

Comment:
`new ConsoleLoggerFactory()` will cause memory leak. `getLogger` method will also cause memory leak, see the implementation in `lib/ConsoleLoggerFactoryImpl.h`:

```c++
Logger* getLogger(const std::string& fileName) { return new SimpleLogger(std::cout, fileName, level_); }
```

The implementation is okay in C++ client because what `getLogger` returns will be passed to the constructor of a thread local `std::unique_ptr` in `lib/LogUtils.h`:

```c++
#define DECLARE_LOG_OBJECT()                                                                     \
    static pulsar::Logger* logger() {                                                            \
        static thread_local std::unique_ptr<pulsar::Logger> threadSpecificLogPtr;                \
        pulsar::Logger* ptr = threadSpecificLogPtr.get();                                        \
        if (PULSAR_UNLIKELY(!ptr)) {                                                             \
            std::string logger = pulsar::LogUtils::getLoggerName(__FILE__);                      \
            threadSpecificLogPtr.reset(pulsar::LogUtils::getLoggerFactory()->getLogger(logger)); \
            ptr = threadSpecificLogPtr.get();                                                    \
        }                                                                                        \
        return ptr;                                                                              \
    }
```

And the what `new ConsoleLoggerFactory` returns will be passed to `ClientConfiguration::setLogger`, which wraps the raw pointer into a `std::shared_ptr`. See `lib/ClientConfiguration.cc`:

```c++
ClientConfiguration& ClientConfiguration::setLogger(LoggerFactory* loggerFactory) {
    impl_->loggerFactory.reset(loggerFactory);
    return *this;
}
```

However, in Python wrapper, only the allocated `LoggerWrapper` will be passed to the `setLogger`.

Therefore, you need to clean up the `fallbackLogger` in the destructor of `LoggerWrapper`. As for the `ConsoleLoggerFactory`, it's better to use a singleton.

Reply:
Thanks @BewareMyPower for the detailed explanation, C++ is not my first language (nor second) so I appreciate it. Can you take a look again? I think I fixed it, although I'm not sure which singleton you were referring to.

Reply:
What I said about singleton is to create a global `ConsoleLoggerFactory` object and reuse it, in C++ it can be implemented simply like:

```c++
inline ConsoleLoggerFactory& globalConsoleLoggerFactory() {
    static ConsoleLoggerFactory factory;  // the object exists during the whole lifetime of the process
    return factory;
}
```

Then you only need to call `globalConsoleLoggerFactory().getLogger("xxx")` to create a logger from the factory. 

But your fix is also OK, you create a `ConsoleLoggerFactory` and delete it every time a `LoggerWrapper` object is constructed and destructed. Since the logger created by it doesn't bind to the factory, your fix is also OK.